text/nodejs/0013-install-bom.md

```diff
@@ -0,0 +1,161 @@
+installation process to the package managers. The case in which we have no
+package manager and use vendored node modules also falls under this case. Due
+to the nature of these dependencies and how they are installed, the process of
+generating a Bill of Materials will require changes to the language family itself.
+
+This is an important piece of the Bill of Materials, because it gives a
+complete picture of what dependencies are inside of the final application
+image. Without this metadata, a major value-add of a Bill of Materials is lost
+and users may have to use third-party tools to generate the rest of the BOM,
+which is a less than ideal user experience.
+
+
+## Implementation
+
+### Tool details
+
+The Bill of Materials that will result should be an accurate accounting of all
+of the modules that get installed into the `node_modules` directory.
+
+There already exists a tool called [CycloneDX Node.js
+Module](https://github.com/CycloneDX/cyclonedx-node-module) which is an
+official CycloneDX supported-tool that creates valid CycloneDX BOMs for Node.JS
+applications. CycloneDX is a BOM format supported by the OWASP Foundation, and
+is a serious contender (at the time of this RFC creation) as a format we will
+eventually support.
+
+The CycloneDX Node.js Module tool can easily be installed via `npm install` and
+then run from a directory containing `node_modules`. The output BOM contains
+the following fields for each module:
+* name,
+* version,
+* description,
+* hash,
+* source URI,
+* license ID,
+* and package URL
+
+All of these are fields that we aim to support per [the Paketo BOM
+RFC](https://github.com/sophiewigmore/rfcs/blob/8b1e8c9ed6201313f47c5897223cbffb265e96ed/text/0028-bill-of-materials.md)
+
+In order to support this tool in both the online and offline cases, we will
+pre-compile the tool with it's dependencies, and provide it as a dependency
+hosted on the [dep-server](https://github.com/paketo-buildpacks/dep-server).
+
+### Language Family Additions
+
+Using the CycloneDX tool to generate the BOM will involve some changes to the
+Node.JS language family buildpack order groups.
+
+* A single BOM generation buildpack will be added. It's primary functions are:
+  * perform the BOM dependency tool retrieval from the dep-server,
+  * execute BOM generation with the tool (`cyclonedx-bom -o bom.json`), and
+  * contribute the BOM to the `launch.toml` `[bom]` label.
+
+* For each of the three Node.JS buildpack order groups, the BOM generation
+  buildpack will run directly before the start command buildpack. This means:
+  * directly before the `yarn-start` buildpack in the Yarn order group,
+  * directly before the `npm-start` buildpack in the NPM order group, and
+  * directly before the `node-start` buildpack in the no-package-manager order group.
+
+#### Detection
+
+The Node Module BOM Generator CNB always detects with the following contract:
+  * Requires {`node`, `node_modules`} during `build`
+  * Provides none
```

Comment:
Seems like it should provide cyclonedx-bom. If we take my sugestion above regarding, shifting responsibility for executing the tool to the buildpack providing the described layer (when appropriate), then
1. if `node_modules` is vendored, it will also require cyclonedx-bom
2. yarn-install and npm-install can require cyclonedx-bom in their first build plan entry (and fall back to passing without generating a bom in a second entry).

Reply:
I think this is correct if we do divide up the generation amongst buildpacks, however I am hesitant to do this fo the reasons mentioned above.